### ISSUE TITLE
Serial broken in mBed OS?

### ISSUE THREAD

**Issue by Timmmm** [ARMmbed/mbed-hal-nrf51822-mcu#44]
I'm having trouble getting serial output to work using `yotta` and mBed OS. Previously it worked. I'm using the nRF51-DK. Here are the steps I'm taking:

1. Install Yotta
2. `mkdir Deleteme`
3. `cd Deleteme`
4. `yt init`
5. (Default answers except 'yes' to executable)
6. `yt target nrf51dk-gcc`
7. `yt install ble`
8. Edit `source/main.cpp`

        #include <mbed-drivers/mbed.h>
        
        Serial pc(USBTX, USBRX);
        DigitalOut led1(LED1);
        
        void app_start(int, char**)
        {
        	for (;;)
        	{
        		pc.puts("Hello World\n");
        		led1 = !led1;
        		wait_ms(200);
        	}
        }

9. `yt build`
10. Flash & reset:

        "C:\Program Files (x86)\Nordic Semiconductor\nrf5x\bin\nrfjprog.exe" --family NRF51 --chiperase --program "build\nrf51dk-gcc\source\deleteme-combined.hex"
        "C:\Program Files (x86)\Nordic Semiconductor\nrf5x\bin\nrfjprog.exe" --family NRF51 --reset

After doing that I connect to it at 9600 baud and sadly don't receive any text. Maybe I am doing something wrong, but this *should* work, right? The LED flashes so I'm pretty the code is running.

**Comment by Timmmm** [comment]
By the way I also tried `RawSerial`; that didn't work either.

**Comment by rgrover** [comment]
callbacks in mbed OS are supposed to be asynchronous; including app_start(). using a wait_ms() or an infinite loop is un-necessary and should be avoided.

Can you just try to do a printf()? it should work. then you can setup periodic callbacks.

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSFW-1916

**Comment by Timmmm** [comment]
I tried this instead but it still doesn't work (it still blinks fine, but no serial output):

	#include <mbed-drivers/mbed.h>

	RawSerial pc(USBTX, USBRX);
	DigitalOut led1(LED1);

	void blinkPrint()
	{
		led1 = !led1;
		printf("Hello World\n");
		pc.printf("Hello PC\n");
	}

	void app_start(int, char**)
	{
		minar::Scheduler::postCallback(blinkPrint).period(minar::milliseconds(500));
	}

By the way I'm sure I read somewhere that if you don't want to use the scheduler you can treat `app_start()` the same as `main()` in mBed 2...

**Comment by rklobo** [comment]
Hi Timm,
Try setting the baud  before calling any printf
pc.baud(115200);

Thanks

**Comment by Timmmm** [comment]
Hmm I have already tried that too. And I tried using different pins and an FTDI cable in case flow control was an issue. Do you guys have an nRF51-DK (or similar) to test with? I may be doing something stupid but I can't think what...

**Comment by jrobeson** [comment]
I have one of these boards and have no problems with the nrf51dk. I don't flash with the nordic tools though ATM, I just copy the file over via the mbed mounted drive.

**Comment by Timmmm** [comment]
Hmm I stand corrected. The FTDI cable actually works - I can read the serial on pin 9, but never get anything through the JLink CDC port. (I tried pin 11 too just in case - I never know which way round TX and RX are!)

May be an issue with the driver, though I did try it on another computer and it also didn't work. And I'm sure it *used* to work. Anyway I guess I'll close this since the nrf51822 HAL code is fine.
